File: julia/Scorio.jl/src/aggregate/prm.jl
"""
    prm_aggregate(step_scores; method="last") -> Float64

Reduce one trace's finite per-step process-reward scores using `"last"`,
`"min"`, `"mean"`, `"prod"`, or `"max"`.
"""
function prm_aggregate(step_scores; method="last")::Float64
    method in ("last", "min", "mean", "prod", "max") ||
        error("method must be one of ('last', 'min', 'mean', 'prod', 'max'); got $method.")
    scores = _flatten_numeric(step_scores, "step_scores")
    isempty(scores) && error("step_scores must be non-empty (L >= 1).")
    all(isfinite, scores) || error("step_scores must all be finite.")
    if method == "last"
        return scores[end]
    elseif method == "min"
        return minimum(scores)
    elseif method == "mean"
        return sum(scores) / length(scores)
    elseif method == "max"
        return maximum(scores)
    end
    return prod(scores)
end

File: js/scorio/src/aggregate/prm.ts
/** Process-reward-model step-score aggregation. */

import {
  asFiniteVector,
  type NumericInput,
} from "./internal/numeric.js";

export type PrmAggregateMethod = "last" | "min" | "mean" | "prod" | "max";

export interface PrmAggregateOptions {
  method?: PrmAggregateMethod;
}

function asStepScores(stepScores: NumericInput): number[] {
  return asFiniteVector(
    stepScores,
    "step_scores",
    "step_scores must be non-empty (L >= 1).",
  );
}

/** Reduce a trace's per-step PRM scores to one reward. */
export function prmAggregate(
  stepScores: NumericInput,
  options: PrmAggregateOptions = {},
): number {
  const method = options.method === undefined ? "last" : options.method;
  if (!(["last", "min", "mean", "prod", "max"] as const).includes(method)) {
    throw new Error(
      `method must be one of last, min, mean, prod, max; got ${String(method)}.`,
    );
  }
  const values = asStepScores(stepScores);
  if (method === "last") return values[values.length - 1]!;

  let total = method === "prod" ? 1 : 0;
  let minimum = Infinity;
  let maximum = -Infinity;
  for (const value of values) {
    total = method === "prod" ? total * value : total + value;
    if (value < minimum) minimum = value;
    if (value > maximum) maximum = value;
  }
  if (method === "min") return minimum;
  if (method === "max") return maximum;
  if (method === "mean") return total / values.length;
  return total;
}
